### DISCUSSION TITLE
Audio issue fix

### DISCUSSION THREAD

**Discussion by sviatoslav-lebediev** [sviatoslav-lebediev/anki-quizlet-importer-extended#156] [Polls]
## AUDIO ISSUE FIX IS PUBLIC NOW, IT SHOULD WORK FOR ALL USERS.

You can always donate here if you want https://www.buymeacoffee.com/moro.programmer 

-----

<img src="https://media3.giphy.com/media/a5viI92PAF89q/giphy.gif"/>

Hi guys, I want to discuss a good way to fix the audio issue. As you know, right now, it's not possible to import audio files from the Quizlet to Anki. You need to select a Skip errors checkbox, and this option will prevent throwing errors, but the import result will not contain audio data. The problem is that Quizlet uses CloudFlare to protect its site; occasionally, CloudFlare blocks requests. We also had this problem with deck data (users could not fetch the HTML page); I've fixed it by adding a fallback proxy call. So right now, if you can't fetch html, the add-on tries to fetch it using my server. It works ok for HTML because it doesn't handle a lot of requests; it needs to do one call per deck. However, adding a server audio proxy will significantly increase the amount of calls/traffic. For example, 1 deck with 100 cards will require 201 requests. I'm ok to pay for the server/proxies that serve HTML, but I don't think I want to spend more money for serving audio :) So, there are a few options that I want to consider as a solution.

1. Implement this on the client side. A user will need to solve the Cloudflare "captcha" and the add-on will need to store cookies, etc. and send it to the Quizlet. I remember that the old add-on version tried to do this, but it made the code more complex. It also will require more time to be implemented :(
2. I will do it on the server side. I've implemented this approach, and it works. But as I've mentioned before, it will cost more money. I can see a few options for dealing with the required money.
	1. I will make this a 'pay as you use' option. Each user will pay some money and receive some credits, for example, 1$ = 500 audio items. This will be a 'fair' option, because if you don't need audio, you will not need to pay at all :) But on the other hand, it will complicate the add-on + I will need to handle all these problems with a payment system.
	2. Somebody will give me a server + proxies :). In this case, nothing will change for users, and if, for some reason, audio doesn't work again, I will have obligations only for this person. If you are ready to do this please send an email at moro.programmer@gmail.com
	3. Some amount of recurring buymeacoffee subscriptions. In this case, I can pay for server/proxies. The add-on will be free for other users as well, so it's not as fair approach as the first one:) One more disadvantage of this approach and the next one for me, is that because I take some money, I will be "obliged" to fix the audio as soon as possible when something happens :(
	4. We collect money one time and keep the add-on free for 1-2-3 years :) Depends on how much we will collect. The advantages and disadvantages are the same as the 3rd one.
3. Somebody else has another idea on how to handle this (add your ideas in the comments)

Please vote in the poll; it will help me to understand what to do next. Feel free to add your comments as well.
p.s. If you are ok with paying more than 5$ once or every month, please vote and add a comment with a sum.

**Comment by malotru007** [comment]
Hi, Thank you for this. I personally would be more in favor of the pay as you use model. Cheers!

**Reply by sviatoslav-lebediev** [reply]
Write me at  moro.programmer@gmail.com and I will send you a key.

**Comment by Nuvolsun** [comment]
I'll be okay with paying in a per use basis. If possible I'd like to know when you'll be implementing the change

**Reply by sviatoslav-lebediev** [reply]
The beta server is almost ready. You can write me at moro.programmer@gmail.com, and I will send you a key. It will be free for now.

**Comment by KasleeV** [comment]
I really need the audio version of the flashcards. I vote to pay as we use as well. Is there any way to facilitate that?

**Reply by sviatoslav-lebediev** [reply]
Write to me at moro.programmer@gmail.com, and I will give you a key.

**Comment by mbryant1832** [comment]
The audio helps a lot with the flashcards I used to study, so I'd be willing to also use Pay as you use.

**Reply by sviatoslav-lebediev** [reply]
thx for the comment; if you want to be a beta tester, write to me at moro.programmer@gmail.com, and I will send you a key

**Comment by aaearon** [comment]
For option 1, is https://github.com/FlareSolverr/FlareSolverr a solution? People hosting their own flaresolverr may be a bit niche and I may not understand the problem with Cloudflare.

**Reply by sviatoslav-lebediev** [reply]
Hi,
Yes, it is a solution, but for cases when we need to solve the challenge. As for me, the better way is to use a big list of proxies and pretend that we are real/good clients; in this way, CloudFlare will not prevent us from downloading data. The proxy approach will also work much faster.

So you understood the problem correctly, but your approach allows us to solve a Cloudflare challenge; my approach allows us to download data without solving any challenges.

**Comment by malotru007** [comment]
Works perfectly!!! Thanks again

**Comment by vaticanskank** [comment]
hi king! would be in favour of a pay per use model. can i get a passkey?

**Comment by vaticanskank** [comment]
Thank you king, I appreciate it

**Comment by sdaqo** [comment]
@sviatoslav-lebediev Hey there, I know this has been stale for a while now, but why can't we source the audio from somewhere else? I may be wrong but quizlet is only tts, right? 

For example the google translate tts:
`https://translate.google.com/translate_tts?ie=UTF-8&q=<here goes the word>&tl=<target language, e.g. en>&client=tw-ob`

**Comment by sviatoslav-lebediev** [comment]
Hi @sdaqo,

I'm not sure if Quizlet is only text-to-speech (TTS); when I used it, the quality was much better than Google TTS. However, you can also record your voice when creating a Quizlet deck, so in this case, TTS will not work. Additionally, you can uncheck the "Download audio" checkbox and use https://ankiweb.net/shared/info/1436550454 if you prefer. 

I've made my implementation free for now, so you no longer need to ask for a key.

P.S. The TTS option can be a good alternative if somebody wants to use a TTS instead, but implementing this feature would require time, which I currently don't have.

**Reply by sdaqo** [reply]
> I'm not sure if Quizlet is only text-to-speech (TTS); when I used it, the quality was much better than Google TTS.

I'm not sure either but I think we can detect it by looking at the audio url: `https://quizlet.com/v1/tts/en.mp3?v=2&b=bnVtYmVycw&s=ZDk2ZjE0YjA`, here we can clearly see that it is tts.

How about giving the user 3 options:
- Audio from quizlet
- TTS Audio (from google or wherever)
- Auto (only download non-tts audio from quizlet rest from somewhere else)

> P.S. The TTS option can be a good alternative if somebody wants to use a TTS instead, but implementing this feature would require time, which I currently don't have.

This is awesome, thank you! But I wonder how this will look in the long run, do you plan to put a price on it? I think having one "backup" option would be good.

> P.S. The TTS option can be a good alternative if somebody wants to use a TTS instead, but implementing this feature would require time, which I currently don't have.

Totally understandable, if you are okay with it, I could look into implementing it.

**Reply by sviatoslav-lebediev** [reply]
`But I wonder how this will look in the long run, do you plan to put a price on it?`
For now, it will be free, from time to time, people donate 5-10$, so I pay for proxies :).


`How about giving the user 3 options:`

Audio from quizlet  - ok, we have this
TTS Audio (from google or wherever) - i guess that's exactly what `AwesomeTTS ` does, and it has a lot of options, so I'm not sure that's easy to implement, maybe we can somehow integrate it with our add-on
Auto (only download non-tts audio from quizlet rest from somewhere else) - I'm ok to add but I'm not sure that people will use this a lot

So, in a nutshell, if you want to add a TTS integration as an option you can definitely do this, just make a pull request and I will merge it. On the other side, I would prefer keeping this add-on simple, less code less errors :) + all these manipulations you can do outside of this add-on

**Comment by toothfairy2025** [comment]
My Anki will not import quizlet. I have deleted Anki twice and the add on. I have looked at the recommendations, and nothing works

**Reply by sviatoslav-lebediev** [reply]
> My Anki will not import quizlet. I have deleted Anki twice and the add on. I have looked at the recommendations, and nothing works

could you please create an issue here https://github.com/sviatoslav-lebediev/anki-quizlet-importer-extended/issues/new/choose
and provide more information ? p.s please send the full error that you receive

**Comment by santi9x** [comment]
Even if i select the skip errors box Anki still encounters an issue, not allowing me to import my quizlets.

**Reply by sviatoslav-lebediev** [reply]
@santi9x could you add more data here https://github.com/sviatoslav-lebediev/anki-quizlet-importer-extended/issues/231 ?
please share your deck url and debug info

**Comment by LiquefiedSpam** [comment]
Reverse cards don't work-- they are the exact same as normal cards

**Reply by sviatoslav-lebediev** [reply]
> Reverse cards don't work-- they are the exact same as normal cards

please let me know if you have changed the Basic Quizlet Extended note type, maybe there is an error
